src/clients/Deluge.ts

```diff
@@ -0,0 +1,236 @@
+		let response;
+		if (!this.loggedIn) {
+			await this.authenticate();
+			response = await this.call({
+				params,
+				method,
+			});
+		} else
+			response = await this.call({
+				params: params,
+				method,
+			});
+
+		return response && response.data && response.data.error === null
+			? response.data
+			: response.data.error;
+	}
+
+	private async call(body: any): Promise<AxiosResponse | null> {
+		return new Promise((resolve) => {
+			body.id = ++this.msgId;
+			if (this.msgId > 1024) this.msgId = 0;
+			axios
+				.post(
+					this.delugeWebUrl.origin + this.delugeWebUrl.pathname,
+					body,
+					{
+						headers: {
+							"Content-Type": "application/json",
+							Cookie: this.delugeCookie,
+						},
+					}
+				)
+				.then((data) => {
+					resolve(data);
+				})
+				.catch(() => resolve(null));
+		});
+	}
+
+	/**
+	 * lists plugins and adds/sets labels
+	 * returns true if successful.
+	 */
+	private async labelEnabled() {
+		const enabledLabels = await this.sendCall(
+			"core.get_enabled_plugins",
+			[]
+		);
+		return enabledLabels["result"].includes("Label");
+	}
+	private async setLabel(torrenthash: string) {
+		if (this.enLabeled) {
+			let setResult = await this.sendCall("label.set_torrent", [
+				torrenthash,
+				this.delugeLabel,
+			]);
+			if (
+				setResult["message"] &&
+				setResult["message"].includes("Unknown Label")
+			) {
+				await this.sendCall("label.add", [this.delugeLabel]);
+				await this.sendCall("label.set_torrent", [
+					torrenthash,
+					this.delugeLabel,
+				]);
+			}
+			return true;
+		} else {
+			return false;
+		}
+	}
+
+	/**
+	 * injects a torrent into deluge client
+	 */
+	async inject(
+		newTorrent: Metafile,
+		searchee: Searchee,
+		path?: string
+	): Promise<InjectionResult> {
+		if (
+			searchee.infoHash &&
+			!(await this.checkCompleted(searchee.infoHash))
+		) {
+			return InjectionResult.TORRENT_NOT_COMPLETE;
+		}
+		const params = this.formatData(
+			`${newTorrent.name}.cross-seed.torrent`,
+			newTorrent.encode().toString("base64"),
+			path
+		);
+		const addResult = await this.sendCall("core.add_torrent_file", params);
+		if (addResult["result"]) {
+			this.setLabel(newTorrent.infoHash);
+			return InjectionResult.SUCCESS;
+		}
+		if (addResult["message"] && addResult["message"].includes("already")) {
+			return InjectionResult.ALREADY_EXISTS;
+		} else {
+			logger.debug({
+				label: Label.DELUGE,
+				message: `injection failed: ${addResult["message"]}`,
+			});
+			return InjectionResult.FAILURE;
+		}
+	}
+
+	/**
+	 * formats the json for rpc calls to inject
+	 */
+	private formatData(filename: string, filedump: string, path: string) {
+		return [
+			filename,
+			filedump,
+			{
+				file_priorities: [],
+				add_paused: false,
+				seed_mode: getRuntimeConfig().skipRecheck,
+				compact_allocation: true,
+				download_location: path,
+				max_connections: -1,
+				max_download_speed: -1,
+				max_upload_slots: -1,
+				max_upload_speed: -1,
```

Comment:
are all of these options required? what happens if we just don't specify them at all?

Reply:
Not specifically, it will default to the client settings, or alternatively when the label is set, that label's settings.

I think seed_mode (skipRecheck) download_location (path) are the only config necessary ones. The rest should be client/label defaults.

Reply:
Let's remove the unnecessary ones

Reply:
Already done in latest commit.